File: code_snippets/ts/src/develop/workflows/submit.ts
import * as restate from "@restatedev/restate-sdk-clients";
import { SignUpWorkflow } from "./signup";
type User = { id: string; email: string };

async function signUpUser(user: User) {
  // <start_submit>
  // import * as restate from "@restatedev/restate-sdk-clients";
  const rs = restate.connect({ url: "http://localhost:8080" });
  const handle = await rs
    .workflowClient<SignUpWorkflow>({ name: "signup" }, "someone")
    .workflowSubmit({ email: user.email });
  // <end_submit>

  // <start_query>
  const status = await rs
    .workflowClient<SignUpWorkflow>({ name: "signup" }, "someone")
    .getStatus();
  // <end_query>

  // <start_attach>
  // Option 1: attach and wait for result with handle
  const result1 = await rs.result(handle);

  // Option 2: attach and wait for result with workflow ID
  const result2 = await rs
    .workflowClient<SignUpWorkflow>({ name: "signup" }, "someone")
    .workflowAttach();

  // Option 3: peek to check if ready with workflow ID
  const peekOutput = await rs
    .workflowClient<SignUpWorkflow>({ name: "signup" }, "someone")
    .workflowOutput();
  if (peekOutput.ready) {
    const result3 = peekOutput.result;
  }
  // <end_attach>
}

File: code_snippets/go/develop/workflows/client/submit.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
)

type User struct {
	Name  string `json:"name"`
	Email string `json:"email"`
}

func MyClientHandler() error {
	// <start_submit>
	const RESTATE_URL = "http://localhost:8080"
	payload, err := json.Marshal(User{
		Name:  "John Doe",
		Email: "john@doe.com",
	})
	if err != nil {
		return err
	}

	url := fmt.Sprintf("%s/SignupWorkflow/someone/Run/send", RESTATE_URL)
	resp, err := http.Post(url, "application/json", bytes.NewBuffer(payload))
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	// <end_submit>
	return nil
}

func MyClientHandler1() error {
	// <start_query>
	const RESTATE_URL = "http://localhost:8080"
	url := fmt.Sprintf("%s/SignupWorkflow/someone/GetStatus", RESTATE_URL)
	resp, err := http.Get(url)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	slog.Info("Response: " + string(body))
	// <end_query>
	return nil
}

func MyClientHandler2() error {
	// <start_attach>
	const RESTATE_URL = "http://localhost:8080"
	attachUrl := fmt.Sprintf("%s/restate/workflow/SignupWorkflow/someone/attach", RESTATE_URL)
	result, err := http.Get(attachUrl)
	if err != nil {
		return err
	}
	defer result.Body.Close()
	// <end_attach>

	return nil
}

func main() {
	if err := MyClientHandler(); err != nil {
		fmt.Println(err)
	}
}
